crates/core/c8y_api/src/http_proxy.rs

```diff
@@ -131,54 +137,107 @@ impl C8yEndPoint {
+    c8y: BasicCredentials,
+}
+
+#[derive(Debug, serde::Deserialize)]
+struct BasicCredentials {
+    username: String,
+    password: String,
 }
 
 #[derive(thiserror::Error, Debug)]
-pub enum JwtRetrieverError {
+pub enum C8yAuthRetrieverError {
     #[error(transparent)]
     MqttConfigBuild(#[from] MqttConfigBuildError),
+
     #[error(transparent)]
     ConfigMulti(#[from] MultiError),
+
+    #[error(transparent)]
+    JwtError(#[from] JwtError),
+
+    #[error(transparent)]
+    InvalidHeaderValue(#[from] InvalidHeaderValue),
+
+    #[error(transparent)]
+    CredentialsFileError(#[from] CredentialsFileError),
 }
 
-impl C8yMqttJwtTokenRetriever {
+impl C8yAuthRetriever {
     pub fn from_tedge_config(
         tedge_config: &TEdgeConfig,
         c8y_profile: Option<&str>,
-    ) -> Result<Self, JwtRetrieverError> {
-        let mqtt_config = tedge_config
-            .mqtt_config()
-            .map_err(MqttConfigBuildError::from)?;
-
-        Ok(Self::new(
-            mqtt_config,
-            tedge_config
-                .c8y
-                .try_get(c8y_profile)?
-                .bridge
-                .topic_prefix
-                .clone(),
-        ))
+    ) -> Result<Self, C8yAuthRetrieverError> {
+        let c8y_config = tedge_config.c8y.try_get(c8y_profile)?;
+        let topic_prefix = c8y_config.bridge.topic_prefix.clone();
+
+        match c8y_config
+            .registration_mode
+            .to_type(&c8y_config.credentials_path)
+        {
+            AuthType::Basic => Ok(Self::Basic {
+                credentials_path: c8y_config.credentials_path.clone(),
+            }),
+            AuthType::Certificate => {
+                let mqtt_config = tedge_config
+                    .mqtt_config()
+                    .map_err(MqttConfigBuildError::from)?;
+
+                let topic = TopicFilter::new_unchecked(&format!("{topic_prefix}/s/dat"));
+                let mqtt_config = mqtt_config
+                    .with_no_session() // Ignore any already published tokens, possibly stale.
+                    .with_subscriptions(topic);
+
+                Ok(Self::Jwt {
+                    mqtt_config: Box::new(mqtt_config),
+                    topic_prefix,
+                })
+            }
+        }
     }
 
-    pub fn new(mqtt_config: mqtt_channel::Config, topic_prefix: TopicPrefix) -> Self {
-        let topic = TopicFilter::new_unchecked(&format!("{topic_prefix}/s/dat"));
-        let mqtt_config = mqtt_config
-            .with_no_session() // Ignore any already published tokens, possibly stale.
-            .with_subscriptions(topic);
-
-        C8yMqttJwtTokenRetriever {
-            mqtt_config,
-            topic_prefix,
-        }
+    pub async fn get_auth_header_value(&mut self) -> Result<HeaderValue, C8yAuthRetrieverError> {
+        let header_value = match &self {
+            Self::Basic { credentials_path } => {
+                let (username, password) = read_c8y_credentials(credentials_path)?;
```

Comment:
Is this a conscious choice to read the password file on each and every HTTP request?

This has the benefit that a wrong password can be corrected on the fly. But this updated password will not be used by the MQTT bridge. So in any case a `tedge reconnect c8y` will be required after a password update.

Reply:
I didn't really think about it deeply. Since we get JWT token in every new HTTP requests, it follows that way. Do you prefer to change it?